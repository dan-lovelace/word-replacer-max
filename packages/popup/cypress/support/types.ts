import { DeepPartial, StorageProvider } from "@worm/types";

export type VisitWithStorageParams = DeepPartial<Record<StorageProvider, any>>;
