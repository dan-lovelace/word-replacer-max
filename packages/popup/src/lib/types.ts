import { VNode } from "preact";

export type PreactChildren = (string | VNode) | (string | VNode)[];
