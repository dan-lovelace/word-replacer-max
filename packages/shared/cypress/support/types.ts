export type TargetProps = Partial<HTMLElement>;

export type VisitMockParams = {
  bodyContents?: string;
  html?: string;
  scriptContents?: string;
  targetContents?: string;
  targetProps?: TargetProps;
  titleContents?: string;
};
