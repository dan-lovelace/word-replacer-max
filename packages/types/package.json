{
  "name": "@worm/types",
  "version": "0.0.0",
  "author": "dan-lovelace",
  "main": "src/index.ts",
  "scripts": {
    "build": "echo \"Info: no build specified\"",
    "test": "echo \"Error: no test specified\" && exit 0"
  },
  "devDependencies": {
    "@types/chrome": "^0.0.287"
  }
}
