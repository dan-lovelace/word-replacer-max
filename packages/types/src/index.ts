import { z } from "zod";

const identificationErrorMessages: Record<IdentificationErrorName, string> = {
  MissingTokens: "Update requires tokens",
  Standard: "Unable to identify user",
  UserNotLoggedIn: "User is not logged in",
};
const queryPatterns = ["case", "default", "regex", "wholeWord"] as const;
const schemaVersions = [1] as const;

export const storageVersions = ["1.0.0", "1.1.0", "1.1.1"] as const;

export type AppUser =
  | {
      email: string;
    }
  | undefined;

export type DeepPartial<T> = T extends object
  ? {
      [P in keyof T]?: DeepPartial<T[P]>;
    }
  : T;

export type DomainEffect = "allow" | "deny";

export type EnvConfigProps = {
  readonly VITE_API_ORIGIN: string;
  readonly VITE_SSM_USER_POOL_HOSTED_UI_QUERY: string;
  readonly VITE_SSM_USER_POOL_CLIENT_ID: string;
  readonly VITE_SSM_USER_POOL_CUSTOM_DOMAIN: string;
  readonly VITE_SSM_USER_POOL_PROVIDER_URL: string;
  readonly VITE_SSM_USER_POOL_ID: string;
  readonly VITE_SSM_USER_POOL_OAUTH_REDIRECT_SIGN_IN: string;
  readonly VITE_SSM_USER_POOL_OAUTH_REDIRECT_SIGN_OUT: string;
  readonly VITE_SSM_USER_POOL_OAUTH_RESPONSE_TYPE: "code" | "token";
  readonly VITE_SSM_USER_POOL_OAUTH_SCOPES: string[];
  readonly VITE_SSM_WEBAPP_ORIGIN: string;
};

export type ExportLink = {
  identifier: ReturnType<Date["getTime"]>;
  url: string;
};

export class IdentificationError extends Error {
  constructor(name: IdentificationErrorName = "Standard") {
    super(identificationErrorMessages[name]);
    this.name = name;

    Object.setPrototypeOf(this, IdentificationError.prototype);
  }
}

type IdentificationErrorName = "MissingTokens" | "Standard" | "UserNotLoggedIn";

export type Matcher = {
  active: boolean;
  identifier: string;
  queries: string[];
  queryPatterns: QueryPattern[];
  replacement: string;

  /**
   * NEW: Introduced during styled replacements, the optional flag should be
   * removed at a later time.
   */
  useGlobalReplacementStyle?: boolean;
};

export type PopupAlertSeverity = "danger" | "info" | "success";

export type PopupTab = "domains" | "options" | "rules" | "support";

export type QueryPattern = (typeof queryPatterns)[number];

export type ReplacementStyle = Partial<{
  active: boolean;
  backgroundColor: string;
  color: string;
  options: ReplacementStyleOption[];
}>;

export type ReplacementStyleOption =
  | "backgroundColor"
  | "bold"
  | "color"
  | "italic"
  | "strikethrough"
  | "underline";

export type SchemaExport = SchemaVersion & {
  version: SchemaVersionType;
};

export type SchemaVersion = SchemaVersion1; // | SchemaVersion2 | SchemaVersion3...

export type SchemaVersionMapper = {
  1: z.ZodType<SchemaVersion1>;
};

export type SchemaVersion1 = {
  version: 1;
  data?: {
    matchers?: Matcher[];
  };
};

export type SchemaVersionType = (typeof schemaVersions)[number];

export type Storage = Partial<{
  [key in StorageKey]: StorageKeyMap[key];
}> & {
  storageVersion?: StorageVersion;
};

export type StorageKey = keyof StorageKeyMap;

export type StorageKeyMap = {
  /**
   * Auth keys are mapped using their respective names from Amplify. A custom
   * token provider exists to translate Amplify's version to our own.
   */
  authAccessToken: string;
  authClockDrift: string;
  authLastAuthUser: string;
  authIdToken: string;
  authRefreshToken: string;

  domainList: string[];
  exportLinks: ExportLink[];
  matchers: Matcher[];
  preferences: {
    activeTab: PopupTab;
    domainListEffect: DomainEffect;
    extensionEnabled: boolean;
    focusRule: Matcher["identifier"];
  };
  replacementStyle: ReplacementStyle;
  storageVersion: StorageVersion;
};

export type StorageSetOptions = {
  onError?: (message: string) => void;
  onSuccess?: () => void;
};

export type StorageVersion = (typeof storageVersions)[number];

export type SystemColor =
  | "black"
  | "blue"
  | "blueGray"
  | "cyan"
  | "green"
  | "orange"
  | "pink"
  | "red"
  | "white"
  | "yellow";

export type WebAppPingResponse = boolean;

export * from "./api";
