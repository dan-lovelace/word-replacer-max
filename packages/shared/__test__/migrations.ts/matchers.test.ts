import {
  translateMatchersForStorage,
  translateStoredMatchers,
} from "@worm/shared/src/migrations";
import { Matcher } from "@worm/types";

describe("translateMatchersForStorage", () => {
  const testMatchers: Matcher[] = [
    {
      active: true,
      identifier: "b7fce47e-58e8-4409-adf4-08da053e713d",
      queries: ["my jaw dropped", "I was shocked"],
      queryPatterns: [],
      replacement: "I was surprised",
    },
    {
      active: true,
      identifier: "34eb8c78-402e-4006-b5a9-b1b15af7a037",
      queries: ["This."],
      queryPatterns: ["case", "wholeWord"],
      replacement: "",
    },
  ];

  it("converts an array of matchers to a shape acceptable for storage", () => {
    expect(translateMatchersForStorage(testMatchers)).toMatchSnapshot();
  });
});

describe("translateStoredMatchers", () => {
  it("converts a raw storage object's matchers to a shape expected by the rest of the system", () => {
    const testStorage = {
      domainList: ["docs.google.com"],
      "matcher__34eb8c78-402e-4006-b5a9-b1b15af7a037": {
        active: true,
        identifier: "34eb8c78-402e-4006-b5a9-b1b15af7a037",
        queries: [],
        queryPatterns: ["case", "wholeWord"],
        replacement: "",
      },
      "matcher__b7fce47e-58e8-4409-adf4-08da053e713d": {
        active: true,
        identifier: "b7fce47e-58e8-4409-adf4-08da053e713d",
        queries: ["my jaw dropped", "I was shocked"],
        queryPatterns: [],
        replacement: "I was surprised",
      },
      preferences: {
        activeTab: "rules",
        domainListEffect: "deny",
        extensionEnabled: true,
      },
    };

    expect(translateStoredMatchers(testStorage)).toMatchSnapshot();
  });

  it("gracefully handles misconfigured stored legacy matchers", () => {
    // create an invalid legacy matcher; notice the missing `identifier` key
    const invalidLegacyMatcher: Omit<Matcher, "identifier"> = {
      active: true,
      queries: ["my jaw dropped", "I was shocked"],
      queryPatterns: [],
      replacement: "I was surprised",
    };

    // also provide a valid legacy matcher
    const validLegacyMatcher: Matcher = {
      active: true,
      identifier: "34eb8c78-402e-4006-b5a9-b1b15af7a037",
      queries: ["This."],
      queryPatterns: ["case", "wholeWord"],
      replacement: "",
    };

    // provide a valid matcher with the new structure to make sure it's preserved
    const newMatcherIdentifier = "1234";
    const newMatcher: Matcher = {
      active: true,
      identifier: newMatcherIdentifier,
      queries: [],
      queryPatterns: ["case", "wholeWord"],
      replacement: "",
    };

    // construct a storage object to test
    const testStorage = {
      matchers: [invalidLegacyMatcher, validLegacyMatcher],
      [`matcher__${newMatcherIdentifier}`]: newMatcher,
    };

    expect(translateStoredMatchers(testStorage)).toMatchSnapshot();
  });
});
