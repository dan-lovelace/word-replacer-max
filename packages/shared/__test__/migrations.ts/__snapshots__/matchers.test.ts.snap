// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translateMatchersForStorage converts an array of matchers to a shape acceptable for storage 1`] = `
{
  "matcher__34eb8c78-402e-4006-b5a9-b1b15af7a037": {
    "active": true,
    "identifier": "34eb8c78-402e-4006-b5a9-b1b15af7a037",
    "queries": [
      "This.",
    ],
    "queryPatterns": [
      "case",
      "wholeWord",
    ],
    "replacement": "",
  },
  "matcher__b7fce47e-58e8-4409-adf4-08da053e713d": {
    "active": true,
    "identifier": "b7fce47e-58e8-4409-adf4-08da053e713d",
    "queries": [
      "my jaw dropped",
      "I was shocked",
    ],
    "queryPatterns": [],
    "replacement": "I was surprised",
  },
}
`;

exports[`translateStoredMatchers converts a raw storage object's matchers to a shape expected by the rest of the system 1`] = `
[
  {
    "active": true,
    "identifier": "34eb8c78-402e-4006-b5a9-b1b15af7a037",
    "queries": [],
    "queryPatterns": [
      "case",
      "wholeWord",
    ],
    "replacement": "",
  },
  {
    "active": true,
    "identifier": "b7fce47e-58e8-4409-adf4-08da053e713d",
    "queries": [
      "my jaw dropped",
      "I was shocked",
    ],
    "queryPatterns": [],
    "replacement": "I was surprised",
  },
]
`;

exports[`translateStoredMatchers gracefully handles misconfigured stored legacy matchers 1`] = `
[
  {
    "active": true,
    "identifier": "1234",
    "queries": [],
    "queryPatterns": [
      "case",
      "wholeWord",
    ],
    "replacement": "",
  },
]
`;
